### ISSUE TITLE
compress filter can panic

### ISSUE THREAD

**Issue by aryszka** [zalando/skipper#1312]
**Describe the bug**
GZip compression may panic sometimes.

**To Reproduce**

1. use a backend that returns uncompressed data, with Content-Type: application/octet-stream 
2. create a route to this backend that includes the compress() filter
3. generate high concurrent load to this route where the requests have the Accept-Encoding: gzip header

**Expected behavior**

compressed responses without problems

**Observed behavior**

rarely, depending on the concurrency level, the following panics are logged from skipper and damaged responses returned:

```
panic: runtime error: index out of range [108] with length 35

goroutine 294066 [running]:
compress/flate.byFreq.Less(...)
	/home/aryszka/go/src/compress/flate/huffman_code.go:339
sort.insertionSort(0x1258840, 0xc00ed08e50, 0x6b, 0x70)
	/home/aryszka/go/src/sort/sort.go:27 +0xc4
sort.quickSort(0x1258840, 0xc00ed08e50, 0x6b, 0x70, 0x9)
	/home/aryszka/go/src/sort/sort.go:209 +0x201
sort.quickSort(0x1258840, 0xc00ed08e50, 0x6b, 0x7c, 0x9)
	/home/aryszka/go/src/sort/sort.go:194 +0xf8
sort.quickSort(0x1258840, 0xc00ed08e50, 0x6b, 0x96, 0xa)
	/home/aryszka/go/src/sort/sort.go:194 +0xf8
sort.Sort(0x1258840, 0xc00ed08e50)
	/home/aryszka/go/src/sort/sort.go:218 +0x79
compress/flate.(*byFreq).sort(...)
	/home/aryszka/go/src/compress/flate/huffman_code.go:333
compress/flate.(*huffmanEncoder).generate(0xc00ed08dc0, 0xc0100e2000, 0x11e, 0x11e, 0xf)
	/home/aryszka/go/src/compress/flate/huffman_code.go:306 +0x188
compress/flate.(*huffmanBitWriter).writeBlockHuff(0xc00ed0ad20, 0xc01012c000, 0xc0100e4000, 0x328, 0xffff)
	/home/aryszka/go/src/compress/flate/huffman_bit_writer.go:639 +0xf7
compress/flate.(*compressor).encSpeed(0xc010038000)
	/home/aryszka/go/src/compress/flate/deflate.go:363 +0x285
compress/flate.(*compressor).syncFlush(0xc010038000, 0x328, 0x2000)
	/home/aryszka/go/src/compress/flate/deflate.go:565 +0x50
compress/flate.(*Writer).Flush(...)
	/home/aryszka/go/src/compress/flate/deflate.go:724
compress/gzip.(*Writer).Flush(0xc000643600, 0xc01a738000, 0x328)
	/home/aryszka/go/src/compress/gzip/gzip.go:221 +0x5f
github.com/zalando/skipper/filters/builtin.encode(0xc033d08cc8, 0x124fa80, 0xc0195d9310, 0x112ca7e, 0x4, 0x1)
	/home/aryszka/work/go/src/github.com/zalando/skipper/filters/builtin/compress.go:341 +0x29e
created by github.com/zalando/skipper/filters/builtin.responseBody
	/home/aryszka/work/go/src/github.com/zalando/skipper/filters/builtin/compress.go:358 +0x1bd
```

**Comment by aryszka** [comment]
investigation shows that the problem is caused by first putting the encoders back to the pool and only then closing them.

**Comment by szuecs** [comment]
@aryszka this can be closed, right?

**Comment by AlexanderYastrebov** [comment]
Fixed by #1313
